keps/sig-node/4603-tune-crashloopbackoff/README.md

```diff
@@ -0,0 +1,1794 @@
+since the Pods have been completely redeployed, they will lose their prior
+backoff counter anyways and, if restarted, will start from the beginning of
+their backoff curve (either the original one with initial value 10s, or the new
+baseline with initial value 1s, depending on whether they've turned on the
+`ReduceDefaultCrashLoopBackoffDecay` feature gate).
+
+Or, the entire cluster can be restarted with the
+`EnableRapidCrashLoopBackoffDecay` feature gate turned off. In this case, any
+Pod configured with `restartPolicy: Rapid` will instead serve as `restartPolicy:
+Always` and use the default backoff curve. Again, since the cluster was
+restarted and Pods were redeployed, they will not maintain prior state and will
+start at the beginning of their backoff curve.
+
+### Version Skew Strategy
+
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and nodes?
+- How does an n-3 kubelet or kube-proxy without this feature available behave when this feature is used?
+- How does an n-1 kube-controller-manager or kube-scheduler without this feature available behave when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+For the default backoff curve, no coordination must be done between the control
+plane and the nodes; all behavior changes are local to the kubelet component and
+its start up configuration.
+
+For the `Rapid` case, the API server and kubelets across all nodes must be
+coordinated:
+
+Firstly, `Rapid` must be a valid option to the `restartPolicy` in the API
+server, which will only be possible if the API server is updated to 1.31. 
+
+Secondly, the `Rapid` value must be interpretable by all kubelets on every node.
+Unfortunately, it is not possible for the API server to be aware of what version
+each kubelet is on, so it cannot serve `Rapid` as `Always` preferentially to
+each kubelet depending on its version. Instead, each kubelet must be able to
+handle this value properly, both at n-3 kubelet version and -- more easily -- at
+its contemporary 1.31+ kubelet version. At 1.31+ kubelet version, each kubelet
+will be able to detect if it has the feature gate on, and if so, interpret
+`Rapid` to use the new rapid backoff curve; and if the feature gate is off,
+interpret it instead as `Always`. But at earlier kubelet versions, `Rapid` must
+be ignored in favor of `Always`. Unfortunately for this KEP, the default value
+for `restartPolicy` is Never, so if kubelet drops unexpected enum values for
+`restartPolicy`, a Pod with `Rapid` will be misconfigured by an old kubelet to
+never restart.
```

Comment:
I'm not sure this is right. Unfortunately, it looks like different code paths use a different default value :anguished: 

- https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/kubelet/container/helpers.go#L105 seems to default to Always
- https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/kubelet/kubelet_pods.go#L1713 defaults to OnFailure
- https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L838-L859 defaults to Always
- https://github.com/kubernetes/kubernetes/blob/a7ca13ea29ba5b3c91fd293cdbaec8fb5b30cee2/pkg/kubelet/status/status_manager.go#L554-L573 defaults to Never

Reply:
Preserved in 005428d555c8cd0191fd643eaac8421465511e8c